xla/service/gpu/cudnn_norm_rewriter.cc

```diff
@@ -672,14 +1048,268 @@ class CudnnNormRewriterVisitor : public DfsHloRewriteVisitor {
+      if (instr->user_count() == 1 &&
+          instr->users()[0]->opcode() == HloOpcode::kConvert &&
+          CompatibleElementType(instr->users()[0])) {
+        instr = instr->users()[0];
+      }
+
+      // Verify the uniqueness of the captured Custom Call and inputs.
+      if (!custom_call || !x || !x0 || !dy || !scale ||
+          std::count(dyk.begin(), dyk.end(), dy) != dyk.size() ||
+          std::count(scalek.begin(), scalek.end(), scale) != scalek.size() ||
+          fused_expectation->unique_id() != fused_expectation0->unique_id() ||
+          fused_norm_factor->unique_id() != fused_norm_factor0->unique_id() ||
+          x->unique_id() != x0->unique_id() ||
+          x_center->unique_id() != x_center0->unique_id()) {
+        VLOG(1) << "Layer norm backward inputs not unique.";
+        return OkStatus();
+      }
+
+      // The captured scalar must be one over the number of elements in the
+      // broadcasted dimensions.
+      float actual_r_nelems = scalar->literal().GetAsDouble({}).value();
+      int64_t nelems = 1;
+      for (int i = 0; i < broadcast->shape().dimensions_size(); ++i) {
+        if (!c_linear_search(broadcast->dimensions(), i)) {
+          nelems *= broadcast->shape().dimensions()[i];
+        }
+      }
+      // The absolute of the difference between the actual scaling factor and
+      // the reference value must not exceed a prescribed threshold.
+      float r_nelems = 1. / static_cast<float>(nelems);
+      float numerical_epsilon = std::numeric_limits<bfloat16>::epsilon();
+      if (!(abs(actual_r_nelems - r_nelems) <
+            ((actual_r_nelems + r_nelems) * numerical_epsilon))) {
+        VLOG(1)
+            << "Layer norm backward broadcast operand outside expected range.";
+        return OkStatus();
+      }
+
+      // Identify Dscale = AddReduce(XCenter * NormFactor * DY)
+      // through a top-down search starting from XCenter.
+      auto find_dscale = [x_center, fused_norm_factor,
+                          dy]() -> HloInstruction* {
+        for (const HloInstruction* multiply_norm_factor : x_center->users()) {
+          multiply_norm_factor =
+              SkipUnaryOps(multiply_norm_factor, /*top-down=*/true);
+          // Verify that the user of XCenter is a multiply with a single user.
+          if (!multiply_norm_factor ||
+              multiply_norm_factor->opcode() != HloOpcode::kMultiply ||
+              multiply_norm_factor->user_count() != 1) {
+            continue;
+          }
+          // Verify that one of the factors is a broadcast operating on the
+          // fused norm factor.
+          bool bcast_operand = multiply_norm_factor->operand(0)->opcode() !=
+                               HloOpcode::kBroadcast;
+          if (multiply_norm_factor->operand(bcast_operand)->opcode() !=
+                  HloOpcode::kBroadcast ||
+              SkipUnaryOps(
+                  SkipUnaryOps(multiply_norm_factor->operand(bcast_operand))
```

Comment:
Why call `SkipUnaryOps` in `SkipUnaryOps(multiply_norm_factor->operand(bcast_operand))`? Isn't `multiply_norm_factor->operand(bcast_operand)` guarenteed to be a broadcast?